test/Test/FS.hs

```diff
@@ -0,0 +1,85 @@
+import           Data.Char (isAsciiLower, isAsciiUpper)
+import qualified Data.List as List
+import qualified Data.Text as Text
+import           System.FS.API
+import qualified System.FS.Sim.MockFS as MockFS
+import           Test.QuickCheck
+import           Test.QuickCheck.Instances ()
+import           Test.Tasty
+import           Test.Tasty.QuickCheck (testProperty)
+import           Test.Util.FS
+
+tests :: TestTree
+tests = testGroup "Test.FS" [
+      -- Simulated file system properties
+      testProperty "prop_propNumOpenHandles" prop_propNumOpenHandles
+    , testProperty "prop_propNoOpenHandles" prop_propNoOpenHandles
+    , testProperty "prop_propNumDirEntries" prop_propNumDirEntries
+    , testProperty "prop_propNoDirEntries" prop_propNoDirEntries
+    ]
+
+{-------------------------------------------------------------------------------
+  Simulated file system properties
+-------------------------------------------------------------------------------}
+
+newtype Path = Path FsPath
+  deriving stock (Show, Eq)
+
+newtype UniqueList a = UniqueList [a]
+  deriving stock Show
+
+instance (Arbitrary a, Eq a) => Arbitrary (UniqueList a) where
+  arbitrary = do
+      xs <- arbitrary
+      pure (UniqueList (List.nub xs))
+  shrink (UniqueList []) = []
+  shrink (UniqueList xs) = UniqueList . List.nub <$> shrink xs
+
+instance Arbitrary Path where
+  arbitrary = Path . mkFsPath . (:[]) <$> ((:) <$> genChar <*> listOf genChar)
+    where
+      genChar = elements (['A'..'Z'] ++ ['a'..'z'])
+  shrink (Path p) = case fsPathToList p of
+      [] -> []
+      t:_ -> [
+          Path p'
+        | t' <- shrink t
+        , let t'' = Text.filter (\c -> isAsciiUpper c || isAsciiLower c) t'
+        , not (Text.null t'')
+        , let p' = fsPathFromList [t']
+        ]
+
+prop_propNumOpenHandles :: UniqueList Path -> Property
+prop_propNumOpenHandles (UniqueList paths) = runSimOrThrow $
+    withSimHasFS prop MockFS.empty $ \hfs _ -> do
+      forM_ paths $ \(Path p) -> hOpen hfs p (WriteMode MustBeNew)
+  where
+    prop = propNumOpenHandles (length paths)
+
+prop_propNoOpenHandles :: Property
+prop_propNoOpenHandles = once $ runSimOrThrow $
+    withSimHasFS propNoOpenHandles MockFS.empty $ \_ _ -> pure ()
+
+prop_propNumDirEntries :: Path -> InfiniteList Bool -> UniqueList Path -> Property
+prop_propNumDirEntries (Path dir) isFiles (UniqueList paths) = runSimOrThrow $
+    withSimHasFS prop MockFS.empty $ \hfs _ -> do
+      createDirectoryIfMissing hfs False dir
+      forM_ (zip (getInfiniteList isFiles) paths) $ \(isFile, Path p) ->
+        if isFile then
+          void $ hOpen hfs (dir </> p) (WriteMode MustBeNew)
+        else
+          createDirectory hfs (dir </> p)
+  where
+    prop = propNumDirEntries dir (length paths)
+
+prop_propNoDirEntries :: Path -> Property
+prop_propNoDirEntries (Path dir) = runSimOrThrow $
+    withSimHasFS (propNoDirEntries dir) MockFS.empty $ \hfs _ -> do
+      createDirectoryIfMissing hfs False dir
```

Comment:
Similarly, is it worth doing the same as above (creating a load of entries) and then deleting them again and checking the dir is still empty?

Reply:
Done